src/dotnet-roslyn-tools/PRTagger/PRTagger.cs

```diff
@@ -2,133 +2,186 @@
-using System.Text;
+using Repository = LibGit2Sharp.Repository;
+
+namespace Microsoft.RoslynTools.PRTagger;
 
-internal class PRTagger
+internal static class PRTagger
 {
+    private const string InsertionLabel = "vs-insertion";
+
     /// <summary>
     /// Creates GitHub issues containing the PRs inserted into a given VS build.
     /// </summary>
     /// <param name="vsBuild">VS build number</param>
     /// <param name="vsCommitSha">Commit SHA for VS build</param>
     /// <param name="settings">Authentication tokens</param>
     /// <param name="logger"></param>
-    /// <param name="cancellationToken"></param>
     /// <returns>Exit code indicating whether issue was successfully created.</returns>
     public static async Task<int> TagPRs(
-        string vsBuild,
-        string vsCommitSha,
+        ImmutableArray<(string vsBuild, string vsCommitSha, string previousVsCommitSha)> vsBuildsAndCommitSha,
         RoslynToolsSettings settings,
-        ILogger logger,
-        CancellationToken cancellationToken)
+        AzDOConnection devdivConnection,
+        HttpClient gitHubClient,
+        ILogger logger)
     {
-        using var devdivConnection = new AzDOConnection(settings.DevDivAzureDevOpsBaseUri, "DevDiv", settings.DevDivAzureDevOpsToken);
         using var dncengConnection = new AzDOConnection(settings.DncEngAzureDevOpsBaseUri, "internal", settings.DncEngAzureDevOpsToken);
-        var connections = new AzDOConnection[] { devdivConnection, dncengConnection };
-
-        // Retrieve VS repo
-        var vsRepository = await GetVSRepositoryAsync(devdivConnection.GitClient);
-
-        // Find previous VS commit SHA
-        var vsCommit = await devdivConnection.GitClient.GetCommitAsync(
-            vsCommitSha, vsRepository.Id, cancellationToken: cancellationToken).ConfigureAwait(false);
-        var previousVsCommitSha = vsCommit.Parents.First();
-
+        // vsBuildsAndCommitSha is ordered from new to old.
+        // For each of the product, check if the product is changed from the newest build, keep creating issues if the product has change.
+        // Stop when
+        // 1. All the VS build has been checked.
+        // 2. If some errors happens.
+        // 3. If we found the issue with the same title has been created. It means the issue is created because the last run of the tagger.
         foreach (var product in VSBranchInfo.AllProducts)
         {
-            // We currently only support creating issues for GitHub repos
-            if (!product.RepoHttpBaseUrl.Contains("github.com"))
+            foreach (var (vsBuild, vsCommitSha, previousVsCommitSha) in vsBuildsAndCommitSha)
             {
-                continue;
+                var result = await TagProductAsync(product, logger, vsCommitSha, vsBuild, previousVsCommitSha, settings, devdivConnection, dncengConnection, gitHubClient).ConfigureAwait(false);
+                if (result is TagResult.Failed or TagResult.IssueAlreadyCreated)
```

Comment:
Is there a goal to tag builds that were missed while the tagger was down? On first run we will tag the first n builds, but then we can never go back further because on subsequent runs we will hit `IssueAlreadyCreated` results that stop us from going back further.

Reply:
I added another `--build`  command, allow you can create based on build. (Like the old tagger does)